File: db4oj/core/src/com/db4o/internal/marshall/ObjectHeader.java
/* Copyright (C) 2004 - 2006  Versant Inc.  http://www.db4o.com */

package com.db4o.internal.marshall;

import com.db4o.*;
import com.db4o.internal.*;


/**
 * @exclude
 */
public final class ObjectHeader {
    
    private final ClassMetadata _classMetadata;
    
    public final MarshallerFamily _marshallerFamily;
    
    public final ObjectHeaderAttributes _headerAttributes;
    
    private int _handlerVersion;
    
    public ObjectHeader(ObjectContainerBase container, ReadWriteBuffer reader){
    	this(container,null,reader);
    }

    public ObjectHeader(ClassMetadata yapClass, ReadWriteBuffer reader){
    	this(null,yapClass,reader);
    }

    public ObjectHeader(StatefulBuffer writer){
        this(writer.container(), writer);
    }
    
    public ObjectHeader(ObjectContainerBase stream, ClassMetadata yc, ReadWriteBuffer reader){
        if (Deploy.debug) {
            reader.readBegin(Const4.YAPOBJECT);
        }
        int classID = reader.readInt();
        _marshallerFamily = readMarshallerFamily(reader, classID);
        
        classID=normalizeID(classID);

        _classMetadata=(yc!=null ? yc : stream.classMetadataForID(classID));

        if (Deploy.debug) {
        	// This check has been added to cope with defragment in debug mode: SlotDefragment#setIdentity()
        	// will trigger calling this constructor with a source db yap class and a target db stream,
        	// thus _yapClass==null. There may be a better solution, since this call is just meant to
        	// skip the object header.
        	if(_classMetadata!=null) {
	        	int ycID = _classMetadata.getID();
		        if (classID != ycID) {
		        	System.out.println("ObjectHeader::init YapClass does not match. Expected ID: " + ycID + " Read ID: " + classID);
		        }
        	}
        }
        _headerAttributes = slotFormat().readHeaderAttributes((ByteArrayBuffer)reader);
    }

    public static ObjectHeader defrag(DefragmentContextImpl context) {
    	ByteArrayBuffer source = context.sourceBuffer();
    	ByteArrayBuffer target = context.targetBuffer();
		ObjectHeader header=new ObjectHeader(context.services().systemTrans().container(),null,source);
    	int newID =context.mapping().mappedID(header.classMetadata().getID());
        if (Deploy.debug) {
            target.readBegin(Const4.YAPOBJECT);
        }
        SlotFormat slotFormat = header.slotFormat();
        slotFormat.writeObjectClassID(target,newID);
        slotFormat.skipMarshallerInfo(target);
        slotFormat.readHeaderAttributes(target);
    	return header;
    }
    
    private SlotFormat slotFormat(){
        return SlotFormat.forHandlerVersion(handlerVersion());
    }
    		
	private MarshallerFamily readMarshallerFamily(ReadWriteBuffer reader, int classID) {
		boolean marshallerAware=marshallerAware(classID);
		_handlerVersion = 0;
        if(marshallerAware) {
            _handlerVersion = reader.readByte();
        }
        MarshallerFamily marshallerFamily=MarshallerFamily.version(_handlerVersion);
		return marshallerFamily;
	}
    
    private boolean marshallerAware(int id) {
    	return id<0;
    }
    
    private int normalizeID(int id) {
    	return (id<0 ? -id : id);
    }

    public ClassMetadata classMetadata() {
        return _classMetadata;
    }
    
    public int handlerVersion() {
        return _handlerVersion;
    }

    public static ObjectHeader scrollBufferToContent(LocalObjectContainer container, ByteArrayBuffer buffer) {
        return new ObjectHeader(container, buffer);
    }
    
}

File: db4on/core/src/core/com/db4o/internal/marshall/ObjectHeader.cs
namespace com.db4o.@internal.marshall
{
	/// <exclude></exclude>
	public sealed class ObjectHeader
	{
		private readonly com.db4o.@internal.ClassMetadata _yapClass;

		public readonly com.db4o.@internal.marshall.MarshallerFamily _marshallerFamily;

		public readonly com.db4o.@internal.marshall.ObjectHeaderAttributes _headerAttributes;

		public ObjectHeader(com.db4o.@internal.ObjectContainerBase stream, com.db4o.@internal.Buffer
			 reader) : this(stream, null, reader)
		{
		}

		public ObjectHeader(com.db4o.@internal.ClassMetadata yapClass, com.db4o.@internal.Buffer
			 reader) : this(null, yapClass, reader)
		{
		}

		public ObjectHeader(com.db4o.@internal.StatefulBuffer writer) : this(writer.GetStream
			(), writer)
		{
		}

		public ObjectHeader(com.db4o.@internal.ObjectContainerBase stream, com.db4o.@internal.ClassMetadata
			 yc, com.db4o.@internal.Buffer reader)
		{
			int classID = reader.ReadInt();
			_marshallerFamily = ReadMarshallerFamily(reader, classID);
			classID = NormalizeID(classID);
			_yapClass = (yc != null ? yc : stream.GetYapClass(classID));
			_headerAttributes = ReadAttributes(_marshallerFamily, reader);
		}

		public static com.db4o.@internal.marshall.ObjectHeader Defrag(com.db4o.@internal.ReaderPair
			 readers)
		{
			com.db4o.@internal.Buffer source = readers.Source();
			com.db4o.@internal.Buffer target = readers.Target();
			com.db4o.@internal.marshall.ObjectHeader header = new com.db4o.@internal.marshall.ObjectHeader
				(readers.Context().SystemTrans().Stream(), null, source);
			int newID = readers.Mapping().MappedID(header.YapClass().GetID());
			header._marshallerFamily._object.WriteObjectClassID(target, newID);
			header._marshallerFamily._object.SkipMarshallerInfo(target);
			ReadAttributes(header._marshallerFamily, target);
			return header;
		}

		public com.db4o.@internal.marshall.ObjectMarshaller ObjectMarshaller()
		{
			return _marshallerFamily._object;
		}

		private com.db4o.@internal.marshall.MarshallerFamily ReadMarshallerFamily(com.db4o.@internal.Buffer
			 reader, int classID)
		{
			bool marshallerAware = MarshallerAware(classID);
			byte marshallerVersion = 0;
			if (marshallerAware)
			{
				marshallerVersion = reader.ReadByte();
			}
			com.db4o.@internal.marshall.MarshallerFamily marshallerFamily = com.db4o.@internal.marshall.MarshallerFamily
				.Version(marshallerVersion);
			return marshallerFamily;
		}

		private static com.db4o.@internal.marshall.ObjectHeaderAttributes ReadAttributes(
			com.db4o.@internal.marshall.MarshallerFamily marshallerFamily, com.db4o.@internal.Buffer
			 reader)
		{
			return marshallerFamily._object.ReadHeaderAttributes(reader);
		}

		private bool MarshallerAware(int id)
		{
			return id < 0;
		}

		private int NormalizeID(int id)
		{
			return (id < 0 ? -id : id);
		}

		public com.db4o.@internal.ClassMetadata YapClass()
		{
			return _yapClass;
		}
	}
}
